Blizzard.IO.RabbitMQ/Implementaions/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using Blizzard.IO.Core;
+using Blizzard.IO.RabbitMQ.Entities;
+using EasyNetQ;
+using EasyNetQ.Topology;
+
+namespace Blizzard.IO.RabbitMQ.Implementaions
+{
+    public class NetqRabbitConsumer<TData> : IConsumer<TData>, IConsumerWithMetadata<TData, RabbitMessageProperties>
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IConcreteTypeDeserializer<TData> _concreteTypeDeserializer;
+        private readonly IQueue _sourceQueue;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        public NetqRabbitConsumer(IBus netqBus, IQueue sourceQueue,
+            IDeserializer<TData> deserializer)
+        {
+            _netqBus = netqBus ?? throw new ArgumentNullException(nameof(netqBus));
+            _sourceQueue = sourceQueue ?? throw new ArgumentNullException(nameof(sourceQueue));
+            _deserializer = deserializer ?? throw new ArgumentNullException(nameof(deserializer));
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, IQueue sourceQueue,
+            IConcreteTypeDeserializer<TData> concreteTypeDeserializer)
+        {
+            _netqBus = netqBus ?? throw new ArgumentNullException(nameof(netqBus));
+            _sourceQueue = sourceQueue ?? throw new ArgumentNullException(nameof(sourceQueue));
+            _concreteTypeDeserializer = concreteTypeDeserializer ?? throw new ArgumentNullException(nameof(concreteTypeDeserializer));
+        }
```

Comment:
you can create another protected ctor that would construct the bus and queue and those 2 public ctors would him (in order to avoid code duplication).

Reply:
f